// wdio-docker-service/lib/utils/colorLogger.js
'use strict';

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const COLORS = {
    INFO: _chalk2.default.hex('#6495ED'),
    ERROR: _chalk2.default.keyword('crimson'),
    WARN: _chalk2.default.keyword('orange'),
    LOG: _chalk2.default.gray
};

const ColorLogger = {
    info(...msg) {
        this._writeLog('info', msg, COLORS.INFO);
    },

    error(...msg) {
        this._writeLog('error', msg, COLORS.ERROR);
    },

    warn(...msg) {
        this._writeLog('warn', msg, COLORS.WARN);
    },

    log(...msg) {
        this._writeLog('log', msg);
    },

    /**
     * @param {String} method Corresponding console method name
     * @param {String} msg Message to log
     * @param {Function} color Chalk color function
     * @private
     */
    _writeLog(method, msg, color = COLORS.LOG) {
        const [first, ...rest] = msg;
        console[method](color(first), ...rest);
    }
};

module.exports = ColorLogger;
